contrib/mesos/pkg/executor/service/service.go

```diff
@@ -85,179 +71,162 @@ func NewKubeletExecutorServer() *KubeletExecutorServer {
-	if err != nil {
-		return err
-	}
+	return nil
+}
 
-	cAdvisorInterface, err := cadvisor.New(s.CAdvisorPort)
-	if err != nil {
-		return err
-	}
+func (s *KubeletExecutorServer) runKubelet(execUpdates <-chan kubelet.PodUpdate, kubeletDone chan<- struct{},
+	staticPodsConfigPath string, apiclient *client.Client) error {
+	kcfg, err := s.KubeletConfig()
+	if err == nil {
+		// apply Messo specific settings
+		kcfg.Builder = func(kc *app.KubeletConfig) (app.KubeletBootstrap, *kconfig.PodConfig, error) {
+			k, pc, err := app.CreateAndInitKubelet(kc)
+			if err != nil {
+				return k, pc, err
+			}
 
-	imageGCPolicy := kubelet.ImageGCPolicy{
-		HighThresholdPercent: s.ImageGCHighThresholdPercent,
-		LowThresholdPercent:  s.ImageGCLowThresholdPercent,
-	}
+			klet := k.(*kubelet.Kubelet)
 
-	diskSpacePolicy := kubelet.DiskSpacePolicy{
-		DockerFreeDiskMB: s.LowDiskSpaceThresholdMB,
-		RootFreeDiskMB:   s.LowDiskSpaceThresholdMB,
-	}
+			s.kletLock.Lock()
+			s.klet = klet
+			s.kletLock.Unlock()
+
+			// decorate kubelet such that it shuts down when the executor is
+			decorated := &executorKubelet{
+				Kubelet:      klet,
+				kubeletDone:  kubeletDone,
+				executorDone: executorDone,
+			}
 
-	//TODO(jdef) intentionally NOT initializing a cloud provider here since:
-	//(a) the kubelet doesn't actually use it
-	//(b) we don't need to create N-kubelet connections to zookeeper for no good reason
-	//cloud := cloudprovider.InitCloudProvider(s.CloudProvider, s.CloudConfigFile)
-	//log.Infof("Successfully initialized cloud provider: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
+			return decorated, pc, nil
+		}
+		kcfg.DockerDaemonContainer = "" // don't move the docker daemon into a cgroup
+		kcfg.Hostname = kcfg.HostnameOverride
+		kcfg.KubeClient = apiclient
+		kcfg.NodeName = kcfg.HostnameOverride
+		kcfg.PodConfig = kconfig.NewPodConfig(kconfig.PodConfigNotificationIncremental, kcfg.Recorder) // override the default pod source
+		kcfg.StandaloneMode = false
+		kcfg.SystemContainer = "" // don't take control over other system processes.
+
+		// create main pod source
+		updates := kcfg.PodConfig.Channel(MESOS_CFG_SOURCE)
+		executorDone := make(chan struct{})
+		go func() {
+			// execUpdates will be closed by the executor on shutdown
+			defer close(executorDone)
 
-	hostNetworkSources, err := kubelet.GetValidatedSources(strings.Split(s.HostNetworkSources, ","))
+			for u := range execUpdates {
+				u.Source = MESOS_CFG_SOURCE
+				updates <- u
+			}
+		}()
+
+		// create static-pods directory file source
+		log.V(2).Infof("initializing static pods source factory, configured at path %q", staticPodsConfigPath)
+		fileSourceUpdates := kcfg.PodConfig.Channel(kubelet.FileSource)
+		kconfig.NewSourceFile(staticPodsConfigPath, kcfg.HostnameOverride, kcfg.FileCheckFrequency, fileSourceUpdates)
+
+		// run the kubelet, until execUpdates is closed
+		err = s.KubeletServer.Run(kcfg)
```

Comment:
I don't understand this comment. I see nothing in the upstream Run() that
checks for nil relating to cloudprovider. What am I missing?

On Wed, Sep 23, 2015 at 10:19 AM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/executor/service/service.go
> <https://github.com/kubernetes/kubernetes/pull/13036#discussion_r40208334>
> :
>
> >
> > -	hostNetworkSources, err := kubelet.GetValidatedSources(strings.Split(s.HostNetworkSources, ","))
> > +			for u := range execUpdates {
> > +				u.Source = MESOS_CFG_SOURCE
> > +				updates <- u
> > +			}
> > +		}()
> > +
> > +		// create static-pods directory file source
> > +		log.V(2).Infof("initializing static pods source factory, configured at path %q", staticPodsConfigPath)
> > +		fileSourceUpdates := kcfg.PodConfig.Channel(kubelet.FileSource)
> > +		kconfig.NewSourceFile(staticPodsConfigPath, kcfg.HostnameOverride, kcfg.FileCheckFrequency, fileSourceUpdates)
> > +
> > +		// run the kubelet, until execUpdates is closed
> > +		err = s.KubeletServer.Run(kcfg)
>
> The cloud provider is only setup if nil is passed to Run.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13036/files#r40208334>.
>